Fixes  pyprobml-review/pyprobml-review#5

nitish1295 opened the pull request:
| Before PR |![image](https://user-images.githubusercontent.com/21181046/163434170-cff3d3da-68d4-44a2-be20-7a76c120e060.png) |
| :-: | :-: |
| After PR | ![image](https://user-images.githubusercontent.com/21181046/163530645-8b3be222-e7e8-44a9-a50d-cf6c149ccb2c.png)|

**Looking for inputs on interactive plots. Please let me know if this notebook needs one.**

karm-patel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,311 @@
+{
```
<p>Add meaningful title in 1st markdown cell, refer textbook for the same</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

nitish1295 replied:
<p>Done</p>

karm-patel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,311 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #13.        latexify(width_scale_factor=2, fig_height=2.0)
</pre><p>What was the problem with fig_height=1.5? can you send me that figures?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

nitish1295 replied:
<p>As discussed pending for discussion</p>

karm-patel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,311 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #97.        n = norm.pdf(x_axis, loc=loc, scale=scale)
</pre><p>I suggest to use meaning ful variable name (PEP8 naming convention) probably you can use.</p><p>n -&gt; norm_pdf</p><p>l-&gt; laplace_pdf </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

nitish1295 replied:
<p>Updated variable names as suggested above</p>

karm-patel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,311 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    _, _ = plot_outlier_effect(save_name="robust_pdf_plot_outliers_latexified.pdf", outliers=[8, 8.75, 9.5])
</pre><p>Ideally, above two plots should have same  y_range to actually know the outlier effect on gaussian distribution. (for example check book figure they have same range 0 to 0.5 for both figures) If they have different height then it is visually difficult to see that gaussian height is decreased.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

nitish1295 replied:
<p>Done, axis are static now with range [0.0,0.5]</p>

karm-patel commented:
Hi @nitish1295, the code looks great!
I have added just a few nits. Also, an interactive plot will be useful, maybe you can add a single slider to change location of 3 outliers, so the reader can visualize the effect of an outlier on given distributions. Also, we should fix y_range in the interactive plot.

karm-patel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,311 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #122.        ax.set_ylabel("$PDF$")
</pre><p>Ideally we should use "pdf" without math symbol ($)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

nitish1295 replied:
<p>Done</p>

nitish1295 commented:
As suggested, below are the images with `fig_height=1.5`. Let me know what you think

![image](https://user-images.githubusercontent.com/21181046/163550791-c2a19bd1-6aac-44d4-901f-fe494c673bf4.png)

I'll accommodate all the other changes once we finalize this

karm-patel commented:
Yes, looks good to me, just a few nits in this
1. decrease legend text size, maybe keep 7?
2. In 1st plot, the gaussian line is not clearly visible.  Probably we can use different linestyle and color for the same.

still, I'll check `fig_height` for this figure with @patel-zeel.

nitish1295 commented:
Hi @karm-patel 

- [ ]  As discussed, I have not changed the `fig_height`.  I have also kept the font size unchanged for now. Pending for further discussion
- [x]  Updated a relevant title for the first markdown column
- [x]  Updated relevant variable names for PDF variables as suggested above
- [x] Updated the y-axis label as `pdf`
- [x] Updated the y-axis to be static from [0.0,0.5] instead of dynamic as per data
- [x] Added an interactive plot that changes the position of the outliers as suggested with a fixed y-axis range

**NOTE:** As per the subtext for figure 2.16(a), it seems like the Gaussian pdf and lapalce pdf overlap each other. I did try and change the color/line style but matplotlib did not display the  gaussian PDF


| Before PR |![image](https://user-images.githubusercontent.com/21181046/163434170-cff3d3da-68d4-44a2-be20-7a76c120e060.png) |
| :-: | :-: |
| After PR | ![image](https://user-images.githubusercontent.com/21181046/163561687-7d4e1b0e-94a8-4a4c-a3f3-150dadbaed6c.png)|

patel-zeel commented:
A small thing for consistency. y_label can be set to $p(x)$.

nitish1295 commented:
Hi @patel-zeel 

Regarding the consistency, I was advised by @karm-patel for the y-axis label. I will update it to $p(x)$

patel-zeel commented:
Sure @nitish1295. I would rather use the word "suggested" than "advised". Also, please remember that you do not always have to follow what I and/or @karm-patel suggest. In case you think there is a better way of doing something, or you are not clear why you should only do this and not that, or there are some issues with our suggestions, feel free to get back to us. 

This is how things work in academia, we help you grow your thinking and we value everyone's opinions :)

nitish1295 commented:
Hi @patel-zeel 

Yes, I understand and will keep that in mind for the future. But since I was new to this and it was my first task. I was following @karm-patel lead on this. :smiley:

patel-zeel commented:
@nitish1295 A few more comments:
* Let us not use a different color for the outliers as it may seem like different data
* We can have a legend entry for data
* Gaussian lines are not visible in the first subplot, even in the original figure. We can fix that by changing the line style. Check [this one](https://github.com/probml/pyprobml/issues/683#issuecomment-1063761706).

nipunbatra commented:
Also, the Laplace PDF is getting cutoff. Perhaps we can choose a higher y upper lim in both a) and b)

The other comments from Karm and Zeel sound very reasonable.

nitish1295 commented:
Hi @patel-zeel,

Initially, I was using a dynamic range for the y-axis as follows:

`y_all_data = jnp.concatenate((norm_PDF, laplace_pdf, studentT_pdf))`
`y_range = [jnp.round(jnp.min(y_all_data), 1), jnp.round(jnp.max(y_all_data), 1)]`

Although this does cause the y-axis for the 2 graphs to have different limits as shown in https://github.com/pyprobml-review/pyprobml-review/pull/8#issue-1205316356. Do you want me to use the above solution or just increase the range of the y-axis?

nitish1295 commented:
For now, I have updated the upper limit for the y-axis to 0.6. Let me know what you think

| Before PR |![image](https://user-images.githubusercontent.com/21181046/163434170-cff3d3da-68d4-44a2-be20-7a76c120e060.png) |
| :-: | :-: |
| After PR |![image](https://user-images.githubusercontent.com/21181046/163996402-2d028802-42e7-441e-aa41-7b25b93475a2.png)|

nitish1295 commented:
Hi @patel-zeel  

- [x]  Fixed font size
- [x]  All data for the bar chart has the same color
- [x]  Created legend entry for data
- [x] Updated the gaussian line, it is visible now
- [x] Updated the y-axis label to $p(x)$
- [x] Added `FIG_DIR` parameter as per new guidelines
- [x]  Removed try/catch for libraries already present in requirements.txt, as per new guidelines
- [x]  Removed .pdf extension while saving image. Images are saved based on the new guidelines.
- [x] Update the y-axis range to accommodate Laplace PDF 
- [x] Removed script dependencies, using probml_utils for utility functions 

The updated images from the code https://github.com/pyprobml-review/pyprobml-review/pull/8#issuecomment-1102543656

patel-zeel commented:
LGTM. Minor suggestions:

* We should not edit any other file on main repo without a fair reason to do so, for example, you have edited `.gitignore` here. Can you remove the changes from it? Also, an off-track question, are you using VS-code as an IDE?
* Can you move the notebook to the relevant folder?

nitish1295 commented:
My bad, I have fixed all that. I did not fetch upstream and submitted it for review. Later realized that the repo was updated and made changes

patel-zeel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,367 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #35.        ss_norm_dist : int, default=30
</pre><p>What is ss?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

nitish1295 replied:
Changed to `samples_norm_dist`, let me know if this works

patel-zeel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,367 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #63.        # Genrate Samples from normal distribution
</pre><p>Fix Typo</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

nitish1295 replied:
Fixed this

nitish1295 replied:
Fixed

patel-zeel commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,367 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #141.        if FIG_DIR and LATEXIFY and len(save_name) &gt; 0:
</pre><p>Again <code>FIG_DIR</code> is not needed here, because <code>FIG_DIR</code> check is inbuilt in <code>savefig</code> command.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

nitish1295 replied:
Removed FIG_DIR check, only LATEXIFY check

nitish1295 replied:
Fixed

nitish1295 commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,367 @@
+{
```
Reverted to only check the LATEXIFY flag

nitish1295 replied:
Fixed

nitish1295 commented on robust_pdf_plot.ipynb:
```diff
@@ -0,0 +1,367 @@
+{
```
Fixed the typo

karm-patel commented:
Hi @nitish1295, thanks for your patience, code looks good to me! let's wait for @patel-zeel 's final approval.

patel-zeel commented:
> Using Jupyter Notebooks(not on VS-Code). I do have VS-code but given that the guidelines mentioned that there could be issues with that, I stayed clear of VS-code

VS-code has a nice pop-up warning you if more than one file is changed. It can be used for git managing.
<img width="36" alt="image" src="https://user-images.githubusercontent.com/59758528/164308293-31ce7600-0fef-457b-b3ff-d342891576c7.png">

The notebook looks great to me!

cc'ing Prof. @nipunbatra for final words.
